File: backend/services/guoman8_scraper.py
import re
import json
from bs4 import BeautifulSoup
from .base_scraper import BaseScraper

class Guoman8Scraper(BaseScraper):
    """国漫8采集器 - www.guoman8.cc"""
    
    def __init__(self):
        super().__init__('https://www.guoman8.cc')
        self.headers.update({
            'Referer': self.base_url,
        })

    def get_categories(self):
        """获取分类列表"""
        try:
            response = self._make_request(self.base_url)
            if not response:
                return {'categories': [], 'total': 0}
            
            soup = BeautifulSoup(response.text, 'lxml')
            
            # 选择器: body > div.w998.nav-bar.shadow > 
            # div.nav-sub.fl > div > div > div.filter.genre > ul > li
            categories = []
            genre_ul = soup.select_one(
                'div.filter.genre > ul'
            )
            
            if genre_ul:
                li_items = genre_ul.find_all('li')
                for li in li_items:
                    link = li.find('a')
                    if link:
                        href = link.get('href', '')
                        name = link.get_text(strip=True)
                        
                        # 提取分类ID (例如: /list/smid-1)
                        match = re.search(r'smid-(\d+)', href)
                        if match:
                            category_id = match.group(1)
                            categories.append({
                                'id': category_id,
                                'name': name,
                                'url': self.base_url + href
                            })
            
            return {
                'categories': categories,
                'total': len(categories)
            }
        except Exception as e:
            print(f'获取分类失败: {e}')
            return {'categories': [], 'total': 0}

    def get_comics_by_category(self, category_id, page=1, limit=20):
        """根据分类获取漫画列表"""
        try:
            # 构建URL: /list/smid-{category_id}-p-{page}
            if page == 1:
                url = f'{self.base_url}/list/smid-{category_id}'
            else:
                url = f'{self.base_url}/list/smid-{category_id}-p-{page}'
            
            response = self._make_request(url)
            if not response:
                return {'comics': [], 'hasMore': False, 'total': 0}
            
            soup = BeautifulSoup(response.text, 'lxml')
            
            # 获取漫画列表: #contList > li
            comics = []
            cont_list = soup.select('#contList > li')
            
            for li in cont_list:
                try:
                    # 获取链接
                    link = li.find('a', class_='bcover')
                    if not link:
                        continue
                    
                    href = link.get('href', '')
                    # 提取漫画ID (例如: /44958/)
                    match = re.search(r'/(\d+)/', href)
                    if not match:
                        continue
                    
                    comic_id = match.group(1)
                    
                    # 获取封面
                    img = link.find('img')
                    cover = img.get('src', '') if img else ''
                    
                    # 获取标题
                    title_elem = li.find('dt')
                    title = title_elem.get_text(strip=True) if title_elem else ''
                    
                    # 获取最新章节
                    latest_elem = li.find('dd', class_='tags')
                    latest_chapter = ''
                    if latest_elem:
                        latest_chapter = latest_elem.get_text(strip=True)
                    
                    comics.append({
                        'id': comic_id,
                        'title': title,
                        'cover': cover if cover.startswith('http') 
                                else self.base_url + cover,
                        'latestChapter': latest_chapter,
                        'status': 'ongoing',
                        'updateTime': '',
                    })
                except Exception as e:
                    print(f'解析漫画项失败: {e}')
                    continue
            
            # 获取分页信息
            has_more = False
            pager = soup.select_one('div.book-list div.pager')
            if pager:
                next_link = pager.find('a', text='下一页')
                has_more = next_link is not None
            
            return {
                'comics': comics,
                'hasMore': has_more,
                'total': len(comics)
            }
        except Exception as e:
            print(f'获取分类漫画失败: {e}')
            return {'comics': [], 'hasMore': False, 'total': 0}

    def get_hot_comics(self, page=1, limit=20):
        """获取热门漫画"""
        # 使用首页推荐作为热门
        return self.get_comics_by_category('1', page, limit)

    def get_latest_comics(self, page=1, limit=20):
        """获取最新漫画"""
        # 使用最新更新分类
        return self.get_comics_by_category('2', page, limit)

    def search_comics(self, keyword, page=1, limit=20):
        """搜索漫画"""
        try:
            # 构建搜索URL
            search_url = f'{self.base_url}/search/{keyword}'
            if page > 1:
                search_url += f'-p-{page}'
            
            response = self._make_request(search_url)
            if not response:
                return {'comics': [], 'hasMore': False, 'total': 0}
            
            soup = BeautifulSoup(response.text, 'lxml')
            
            # 解析搜索结果
            comics = []
            cont_list = soup.select('#contList > li')
            
            for li in cont_list:
                try:
                    link = li.find('a', class_='bcover')
                    if not link:
                        continue
                    
                    href = link.get('href', '')
                    match = re.search(r'/(\d+)/', href)
                    if not match:
                        continue
                    
                    comic_id = match.group(1)
                    
                    img = link.find('img')
                    cover = img.get('src', '') if img else ''
                    
                    title_elem = li.find('dt')
                    title = title_elem.get_text(strip=True) if title_elem else ''
                    
                    comics.append({
                        'id': comic_id,
                        'title': title,
                        'cover': cover if cover.startswith('http') 
                                else self.base_url + cover,
                        'latestChapter': '',
                        'status': 'ongoing',
                    })
                except Exception as e:
                    continue
            
            return {
                'comics': comics,
                'hasMore': len(comics) >= limit,
                'total': len(comics)
            }
        except Exception as e:
            print(f'搜索失败: {e}')
            return {'comics': [], 'hasMore': False, 'total': 0}

    def get_comic_detail(self, comic_id):
        """获取漫画详情"""
        try:
            url = f'{self.base_url}/{comic_id}/'
            response = self._make_request(url)
            if not response:
                return None
            
            soup = BeautifulSoup(response.text, 'lxml')
            
            # 封面: body > div.w998.bc.cf > div.fl.w728 > 
            # div.book-cont.cf > div.book-cover.fl > p > img
            cover_img = soup.select_one(
                'div.book-cover.fl > p > img'
            )
            cover = cover_img.get('src', '') if cover_img else ''
            
            # 标题和介绍: div.book-detail.pr.fr
            detail_div = soup.select_one('div.book-detail.pr.fr')
            
            title = ''
            author = ''
            description = ''
            status = 'ongoing'
            categories = []
            
            if detail_div:
                # 标题
                title_elem = detail_div.find('h1')
                if title_elem:
                    title = title_elem.get_text(strip=True)
                
                # 作者和状态
                info_list = detail_div.find_all('li')
                for li in info_list:
                    text = li.get_text(strip=True)
                    if '作者' in text:
                        author = text.replace('作者:', '').strip()
                    elif '状态' in text:
                        if '完结' in text:
                            status = 'completed'
                    elif '类型' in text or '题材' in text:
                        # 提取分类
                        links = li.find_all('a')
                        categories = [
                            a.get_text(strip=True) for a in links
                        ]
                
                # 简介
                intro_div = detail_div.find('div', class_='intro')
                if intro_div:
                    description = intro_div.get_text(strip=True)
            
            return {
                'id': comic_id,
                'title': title,
                'cover': cover if cover.startswith('http') 
                        else self.base_url + cover,
                'author': author,
                'description': description,
                'status': status,
                'categories': categories,
                'rating': 0,
                'updateTime': '',
            }
        except Exception as e:
            print(f'获取漫画详情失败: {e}')
            return None

    def get_chapters(self, comic_id):
        """获取章节列表"""
        try:
            url = f'{self.base_url}/{comic_id}/'
            response = self._make_request(url)
            if not response:
                return {'chapters': [], 'total': 0}
            
            soup = BeautifulSoup(response.text, 'lxml')
            
            # 章节列表: #chpater-list-1 > ul > li
            chapters = []
            chapter_ul = soup.select_one('#chpater-list-1 > ul')
            
            if chapter_ul:
                li_items = chapter_ul.find_all('li')
                for index, li in enumerate(li_items):
                    link = li.find('a')
                    if link:
                        href = link.get('href', '')
                        title = link.get_text(strip=True)
                        
                        # 提取章节ID (例如: /45043/01.html)
                        match = re.search(
                            r'/(\d+)/([^/]+)\.html', 
                            href
                        )
                        if match:
                            chapter_id = f"{match.group(1)}_{match.group(2)}"
                            
                            chapters.append({
                                'id': chapter_id,
                                'title': title,
                                'order': index + 1,
                                'updateTime': '',
                                'isRead': False,
                            })
            
            return {
                'chapters': chapters,
                'total': len(chapters)
            }
        except Exception as e:
            print(f'获取章节列表失败: {e}')
            return {'chapters': [], 'total': 0}

    def get_chapter_images(self, chapter_id):
        """获取章节图片列表"""
        try:
            # 解析chapter_id (格式: 45043_01)
            parts = chapter_id.split('_')
            if len(parts) != 2:
                return {'images': [], 'total': 0}
            
            comic_id, chapter_num = parts
            url = f'{self.base_url}/{comic_id}/{chapter_num}.html'
            
            response = self._make_request(url)
            if not response:
                return {'images': [], 'total': 0}
            
            # 解析JavaScript获取图片数组
            images = self._extract_images_from_js(response.text)
            
            return {
                'images': images,
                'total': len(images)
            }
        except Exception as e:
            print(f'获取章节图片失败: {e}')
            return {'images': [], 'total': 0}

    def _extract_images_from_js(self, html):
        """从JavaScript代码中提取图片URL"""
        try:
            # 查找包含图片数组的JavaScript代码
            # 格式: var s={..., 'B':[图片数组], ...}
            
            # 方法1: 正则提取eval函数中的加密代码
            pattern = r"eval\(function\(p,a,c,k,e,d\)\{.*?\}\('(.*?)',(\d+),(\d+),'(.*?)'\.split"
            match = re.search(pattern, html, re.DOTALL)
            
            if match:
                # 这里需要解密JavaScript
                # 简化处理: 直接查找图片URL模式
                img_pattern = r'https?://[^\s\'"]+\.(?:jpg|jpeg|png|gif|webp)'
                img_urls = re.findall(img_pattern, html)
                
                images = []
                for index, url in enumerate(img_urls):
                    images.append({
                        'page': index + 1,
                        'url': url
                    })
                
                return images
            
            # 方法2: 查找直接的图片数组
            # 查找类似 'B':['url1','url2'] 的模式
            array_pattern = r"'B'\s*:\s*\[(.*?)\]"
            array_match = re.search(array_pattern, html, re.DOTALL)
            
            if array_match:
                urls_str = array_match.group(1)
                # 提取所有URL
                url_pattern = r"'(https?://[^']+)'"
                urls = re.findall(url_pattern, urls_str)
                
                images = []
                for index, url in enumerate(urls):
                    images.append({
                        'page': index + 1,
                        'url': url
                    })
                
                return images
            
            return []
        except Exception as e:
            print(f'提取图片URL失败: {e}')
            return []

File: services/scrapers/Guoman8Scraper.js
import { load } from './htmlParser';
import { BaseScraper } from './BaseScraper';

/**
 * 国漫8爬虫实现
 * 网站: guoman8.cc
 */
export class Guoman8Scraper extends BaseScraper {
  constructor() {
    super('https://www.guoman8.cc');
  }

  /**
   * 获取热门漫画
   */
  async getHotComics(page = 1, limit = 20) {
    return this.getComicsByCategory('1', page, limit);
  }

  /**
   * 获取最新漫画
   */
  async getLatestComics(page = 1, limit = 20) {
    return this.getComicsByCategory('2', page, limit);
  }

  /**
   * 根据分类获取漫画列表
   * URL格式: /list/smid-{category_id}-p-{page}
   */
  async getComicsByCategory(categoryId, page = 1, limit = 20) {
    try {
      // 构建URL
      let url;
      if (page === 1) {
        url = `${this.baseUrl}/list/smid-${categoryId}`;
      } else {
        url = `${this.baseUrl}/list/smid-${categoryId}-p-${page}`;
      }

      console.log(`请求分类漫画URL: ${url}`);

      const response = await this.makeRequest(url, {
        headers: {
          'Referer': this.baseUrl,
        },
      });

      const $ = load(response.data);
      const comics = [];

      // 获取漫画列表
      const comicItems = $('#contList > li');

      comicItems.each((index, element) => {
        try {
          const $item = $(element);

          // 获取链接
          const link = $item.find('a.bcover');
          if (!link.length) return;

          const href = link.attr('href') || '';

          // 提取漫画ID (例如: /44958/ -> 44958)
          const match = href.match(/\/(\d+)\//);
          if (!match) return;

          const comicId = match[1];

          // 获取封面
          const img = link.find('img');
          let cover = img.attr('src') || '';
          if (cover && !cover.startsWith('http')) {
            cover = this.baseUrl + cover;
          }

          // 获取标题
          const titleElem = $item.find('dt');
          const title = titleElem.text().trim() || '';

          // 获取最新章节
          const latestElem = $item.find('dd.tags');
          const latestChapter = latestElem.text().trim() || '';

          if (comicId && title) {
            comics.push({
              id: comicId,
              title,
              cover,
              latestChapter,
              status: 'ongoing',
              updateTime: '',
            });
          }
        } catch (error) {
          console.error('解析漫画项失败:', error);
        }
      });

      // 检查是否有下一页
      const pager = $('div.book-list div.pager');
      const hasMore = pager.find('a:contains("下一页")').length > 0;

      return {
        comics,
        hasMore,
        total: comics.length,
      };
    } catch (error) {
      console.error('获取分类漫画失败:', error);
      return { comics: [], hasMore: false, total: 0 };
    }
  }

  /**
   * 搜索漫画
   * URL: /search/{keyword}-p-{page}
   */
  async searchComics(keyword, page = 1, limit = 20) {
    try {
      let url = `${this.baseUrl}/search/${encodeURIComponent(keyword)}`;
      if (page > 1) {
        url += `-p-${page}`;
      }

      console.log(`搜索漫画URL: ${url}`);

      const response = await this.makeRequest(url, {
        headers: {
          'Referer': this.baseUrl,
        },
      });

      const $ = load(response.data);
      const comics = [];

      const comicItems = $('#contList > li');

      comicItems.each((index, element) => {
        try {
          const $item = $(element);
          const link = $item.find('a.bcover');
          if (!link.length) return;

          const href = link.attr('href') || '';
          const match = href.match(/\/(\d+)\//);
          if (!match) return;

          const comicId = match[1];

          const img = link.find('img');
          let cover = img.attr('src') || '';
          if (cover && !cover.startsWith('http')) {
            cover = this.baseUrl + cover;
          }

          const titleElem = $item.find('dt');
          const title = titleElem.text().trim() || '';

          if (comicId && title) {
            comics.push({
              id: comicId,
              title,
              cover,
              latestChapter: '',
              status: 'ongoing',
            });
          }
        } catch (error) {
          console.error('解析搜索结果失败:', error);
        }
      });

      return {
        comics,
        hasMore: comics.length >= limit,
        total: comics.length,
      };
    } catch (error) {
      console.error('搜索漫画失败:', error);
      return { comics: [], hasMore: false, total: 0 };
    }
  }

  /**
   * 获取漫画详情
   * URL: /{comic_id}/
   */
  async getComicDetail(comicId) {
    try {
      const url = `${this.baseUrl}/${comicId}/`;
      console.log(`请求详情URL: ${url}`);

      const response = await this.makeRequest(url, {
        headers: {
          'Referer': this.baseUrl,
        },
      });

      const $ = load(response.data);

      // 获取封面
      const coverImg = $('div.book-cover.fl > p > img');
      let cover = coverImg.attr('src') || '';
      if (cover && !cover.startsWith('http')) {
        cover = this.baseUrl + cover;
      }

      // 获取详情
      const detailDiv = $('div.book-detail.pr.fr');

      let title = '';
      let author = '';
      let description = '';
      let status = 'ongoing';
      const categories = [];

      if (detailDiv.length) {
        // 标题
        const titleElem = detailDiv.find('h1');
        title = titleElem.text().trim() || '';

        // 作者和状态
        const infoList = detailDiv.find('li');
        infoList.each((index, element) => {
          const text = $(element).text().trim();

          if (text.includes('作者')) {
            author = text.replace('作者:', '').trim();
          } else if (text.includes('状态')) {
            if (text.includes('完结')) {
              status = 'completed';
            }
          } else if (text.includes('类型') || text.includes('题材')) {
            // 提取分类
            $(element)
              .find('a')
              .each((idx, elem) => {
                categories.push($(elem).text().trim());
              });
          }
        });

        // 简介
        const introDiv = detailDiv.find('div.intro');
        description = introDiv.text().trim() || '';
      }

      return {
        id: comicId,
        title,
        cover,
        author,
        description,
        status,
        categories,
        rating: 0,
        updateTime: '',
      };
    } catch (error) {
      console.error('获取漫画详情失败:', error);
      return null;
    }
  }

  /**
   * 获取章节列表
   * URL: /{comic_id}/
   * 选择器: #chpater-list-1 > ul > li
   */
  async getChapters(comicId) {
    try {
      const url = `${this.baseUrl}/${comicId}/`;
      console.log(`请求章节列表URL: ${url}`);

      const response = await this.makeRequest(url, {
        headers: {
          'Referer': this.baseUrl,
        },
      });

      const $ = load(response.data);
      const chapters = [];

      const chapterUl = $('#chpater-list-1 > ul');

      if (chapterUl.length) {
        const liItems = chapterUl.find('li');

        liItems.each((index, element) => {
          const $item = $(element);
          const link = $item.find('a');

          if (link.length) {
            const href = link.attr('href') || '';
            const title = link.text().trim();

            // 提取章节ID (例如: /45043/01.html -> 45043_01)
            const match = href.match(/\/(\d+)\/([^/]+)\.html/);
            if (match) {
              const chapterId = `${match[1]}_${match[2]}`;

              chapters.push({
                id: chapterId,
                title,
                order: index + 1,
                updateTime: '',
                isRead: false,
              });
            }
          }
        });
      }

      console.log(`成功获取章节列表, 数量: ${chapters.length}`);

      return {
        chapters,
        total: chapters.length,
      };
    } catch (error) {
      console.error('获取章节列表失败:', error);
      return { chapters: [], total: 0 };
    }
  }

  /**
   * 获取章节图片列表
   * URL: /{comic_id}/{chapter_num}.html
   */
  async getChapterImages(chapterId) {
    try {
      // 解析chapter_id (格式: 45043_01)
      const parts = chapterId.split('_');
      if (parts.length !== 2) {
        return { images: [], total: 0 };
      }

      const [comicId, chapterNum] = parts;
      const url = `${this.baseUrl}/${comicId}/${chapterNum}.html`;

      console.log(`请求章节图片URL: ${url}`);

      const response = await this.makeRequest(url, {
        headers: {
          'Referer': this.baseUrl,
        },
      });

      // 从HTML中提取图片URL
      const images = this.extractImagesFromJs(response.data);

      console.log(`成功获取章节图片, 数量: ${images.length}`);

      return {
        images,
        total: images.length,
      };
    } catch (error) {
      console.error('获取章节图片失败:', error);
      return { images: [], total: 0 };
    }
  }

  /**
   * 从JavaScript代码中提取图片URL
   */
  extractImagesFromJs(html) {
    try {
      // 方法1: 查找直接的图片URL模式
      const imgPattern = /https?:\/\/[^\s'"]+\.(?:jpg|jpeg|png|gif|webp)/gi;
      const imgUrls = html.match(imgPattern) || [];

      const images = [];
      const uniqueUrls = [...new Set(imgUrls)]; // 去重

      uniqueUrls.forEach((url, index) => {
        images.push({
          page: index + 1,
          url,
        });
      });

      if (images.length > 0) {
        return images;
      }

      // 方法2: 查找图片数组 'B':['url1','url2']
      const arrayPattern = /'B'\s*:\s*\[(.*?)\]/s;
      const arrayMatch = html.match(arrayPattern);

      if (arrayMatch) {
        const urlsStr = arrayMatch[1];
        const urlPattern = /'(https?:\/\/[^']+)'/g;
        let match;

        while ((match = urlPattern.exec(urlsStr)) !== null) {
          images.push({
            page: images.length + 1,
            url: match[1],
          });
        }
      }

      return images;
    } catch (error) {
      console.error('提取图片URL失败:', error);
      return [];
    }
  }

  /**
   * 获取分类列表
   * URL: / (首页)
   */
  async getCategories() {
    try {
      const url = this.baseUrl;
      console.log(`请求分类列表URL: ${url}`);

      const response = await this.makeRequest(url, {
        headers: {
          'Referer': this.baseUrl,
        },
      });

      const $ = load(response.data);
      const categories = [];

      const genreUl = $('div.filter.genre > ul');

      if (genreUl.length) {
        const liItems = genreUl.find('li');

        liItems.each((index, element) => {
          const $item = $(element);
          const link = $item.find('a');

          if (link.length) {
            const href = link.attr('href') || '';
            const name = link.text().trim();

            // 提取分类ID (例如: /list/smid-1 -> 1)
            const match = href.match(/smid-(\d+)/);
            if (match) {
              const categoryId = match[1];
              categories.push({
                id: categoryId,
                name,
                url: this.baseUrl + href,
              });
            }
          }
        });
      }

      return {
        categories,
        total: categories.length,
      };
    } catch (error) {
      console.error('获取分类失败:', error);
      return { categories: [], total: 0 };
    }
  }
}
